File: frontend/src/lib/git.ts
import {
  GitBlame,
  GitBranches,
  GitCheckout,
  GitPull,
  GitPush,
  GitRemotes,
  GitStatus,
} from "../../wailsjs/go/app/App";

/**
 * The frontend's view of `internal/git`'s status types.
 *
 * These are declared here rather than aliased from `wailsjs/go/models` — the
 * convention `tree.ts`'s `Entry` and `files.ts`'s `FileContent` follow —
 * because the generated `git.Status` is a class with a method on it, so no
 * object literal is assignable to it. That would push every test fixture and
 * every empty value through the generator's untyped `createFrom`, which takes
 * `any` and would let a wrong shape through silently.
 *
 * The check the alias would have given is kept instead by `wailsGit` below:
 * assigning the generated binding to `Git` fails to compile if the Go structs
 * stop matching these declarations.
 */
export interface Branch {
  readonly name: string;
  readonly upstream: string;
  readonly ahead: number;
  readonly behind: number;
  readonly detached: boolean;
  readonly unborn: boolean;
}

export interface FileStatus {
  readonly path: string;
  /** How the index differs from HEAD; `""` when it does not. */
  readonly staged: string;
  /** How the working tree differs from the index; `""` when it does not. */
  readonly worktree: string;
  readonly conflicted: boolean;
  /** Where a renamed or copied path came from; `""` otherwise. */
  readonly origPath: string;
}

export interface Status {
  readonly availability: string;
  readonly branch: Branch;
  readonly files: readonly FileStatus[];
}

/** One commit a file's lines are attributed to (#52), matching
 * `internal/git.BlameCommit`. */
export interface BlameCommit {
  readonly sha: string;
  readonly author: string;
  /** When the author wrote it, in Unix seconds; 0 when git reported no
   * readable time. */
  readonly authorTime: number;
  /** The commit's subject line. */
  readonly summary: string;
  /** git's all-zero SHA: work in the working tree and in no commit. Such a
   * commit has no author or date worth showing. */
  readonly uncommitted: boolean;
}

/**
 * One file's per-line attribution, in the shape git's porcelain format states
 * it: each commit once, and a line-by-line reference into them.
 *
 * `lines` holds one index into `commits` per line of the file, line 1 first.
 * An index outside `commits` means the blame did not cover that line — see
 * `commitAt` in `blame.ts`, which is the only thing that should read this.
 */
export interface Blame {
  readonly commits: readonly BlameCommit[];
  readonly lines: readonly number[];
}

/**
 * Why a project has no readable git state, matching
 * `internal/git.Availability`.
 *
 * These are constants rather than a union type because Wails erases Go's
 * named string types to `string` in the generated models: a union here would
 * make the binding unassignable to `Git` and cost the compile-time check that
 * is the whole reason the interfaces above are hand-written.
 */
export const AVAILABLE = "ok";
export const NO_GIT = "no-git";
export const NOT_A_REPOSITORY = "not-a-repository";

/** Per-path change kinds, matching `internal/git.State`. */
export const MODIFIED = "modified";
export const ADDED = "added";
export const DELETED = "deleted";
export const RENAMED = "renamed";
export const COPIED = "copied";
export const UNTRACKED = "untracked";

/**
 * The git service's Wails-binding seam (DESIGN.md §7), the same shape
 * `lib/directory.ts`'s `Directory` already takes: an interface a component or
 * hook can be tested against without a Wails runtime.
 *
 * Only `status` and the two list calls answer with anything. The mutations
 * resolve with nothing on success and reject with git's own message on
 * failure, because the backend does not return a status from an operation:
 * the write changes `.git`, the watcher publishes a `git` event for it
 * (PROTOCOL.md §5), and the status is read back through `status`. One source
 * for what the repository looks like, not two that can disagree.
 *
 * Staging and committing are absent, and the absence is the decision (#39):
 * what records work in m6t is the agent in the terminal below, running the
 * user's own git in the user's own worktree. A seam method for it would be a
 * second writer of the index that the agent cannot see.
 */
export interface Git {
  status: (root: string) => Promise<Status>;
  /** One file's per-line attribution (#52). `path` is root-relative and
   * slash-separated; a path git will not blame rejects with git's own words
   * rather than resolving to an empty blame. */
  blame: (root: string, path: string) => Promise<Blame>;
  pull: (root: string) => Promise<void>;
  /** `remote` takes effect only when `setUpstream` is true; otherwise the
   * repository's own push configuration decides where the branch goes. */
  push: (root: string, remote: string, setUpstream: boolean) => Promise<void>;
  checkout: (root: string, branch: string) => Promise<void>;
  branches: (root: string) => Promise<string[]>;
  remotes: (root: string) => Promise<string[]>;
}

/** The git seam backed by the generated Wails bindings. */
export const wailsGit: Git = {
  status: (root) => GitStatus(root),
  blame: (root, path) => GitBlame(root, path),
  pull: (root) => GitPull(root),
  push: (root, remote, setUpstream) => GitPush(root, remote, setUpstream),
  checkout: (root, branch) => GitCheckout(root, branch),
  branches: (root) => GitBranches(root),
  remotes: (root) => GitRemotes(root),
};

/** A blame with nothing in it — a file of no lines. Every consumer already
 * handles a line the blame does not cover, so this needs no special case. */
export function emptyBlame(): Blame {
  return { commits: [], lines: [] };
}

/** A status for a project nothing has been read for yet: available-shaped and
 * empty, so every consumer renders it without a null check. */
export function emptyStatus(): Status {
  return {
    availability: AVAILABLE,
    branch: { name: "", upstream: "", ahead: 0, behind: 0, detached: false, unborn: false },
    files: [],
  };
}

File: internal/git/git.go
// Package git reports what the system `git` says about a project's working
// tree (DESIGN.md §3.2, §7): which paths differ from the index and from HEAD,
// which are unmerged, and where the current branch sits relative to its
// upstream.
//
// It is mostly a reader: status and blame mutate nothing, so every invocation
// of them is safe to run on a filesystem event, which is how the workbench
// keeps its badges current. ops.go is the other half of DESIGN.md §7 — pull,
// push and branch switch — and it goes through the same runner rather than a
// second one of its own.
//
// That runner is internal/gitexec, which is where the process is actually
// started: it pins the locale the not-a-repository match depends on, keeps a
// read from rewriting .git/index, bounds the call, and carries git's stderr out
// verbatim. This package parses what comes back and knows nothing else about
// how it got there. The split exists so a second git reader can be written
// without duplicating any of that (#53); the parsers here are not it.
//
// When git is missing, or the path is no longer a repository, that is reported
// as a state on Status rather than as an error: both are things a user can see
// and fix, and neither is a failure of the call. Anything else — a corrupt
// index, a permission error — is an error carrying git's own stderr verbatim,
// because the user knows how to read a git error and a translation would only
// lose detail (DESIGN.md §7).
//
// The package knows nothing about projects, the Wails bridge or the stream
// transport. It takes a worktree path and returns a value; internal/app is
// what binds that to a project and pushes the change notification that makes
// the UI ask again.
package git

// State classifies how one path differs, on one side of the index.
//
// The values are the wire form: they cross the Wails bridge into TypeScript
// and are compared there, so renaming one is a frontend change too. The empty
// string means "no difference on this side", which is what lets a caller test
// a side with `!= ""` rather than carrying a fourth "unmodified" constant that
// every switch would have to remember to handle.
type State string

const (
	// StateModified is content that differs. A typechange (a file becoming a
	// symlink, git's `T`) is reported as this: the tree shows one badge per
	// path, and "the content is not what it was" is what the badge means.
	StateModified State = "modified"

	// StateAdded is a path present on this side but not the one before it.
	StateAdded State = "added"

	// StateDeleted is a path gone from this side.
	StateDeleted State = "deleted"

	// StateRenamed is a path git matched to a different name in HEAD. The
	// name it came from is FileStatus.OrigPath.
	StateRenamed State = "renamed"

	// StateCopied is a path git matched to a source it was copied from,
	// also carried in FileStatus.OrigPath. git only detects copies when
	// asked to, so this is rare in practice and is here because the format
	// can produce it, not because m6t requests it.
	StateCopied State = "copied"

	// StateUntracked is a path git is not tracking. It is only ever a
	// worktree-side state: an untracked path has nothing in the index to
	// compare against.
	StateUntracked State = "untracked"
)

// Availability says whether a project has readable git state, and if not, why.
//
// It is a field rather than an error because both failure values are ordinary
// conditions with a clear thing for the UI to say. A user with no git
// installed, or a project whose checkout has stopped being a repository,
// should see a sentence in the status bar — not an error box every time a file
// is saved.
type Availability string

const (
	// Available means git ran and the status below it is real.
	Available Availability = "ok"

	// NoGit means no `git` executable was found on PATH. Every other field
	// of the Status is zero.
	NoGit Availability = "no-git"

	// NotARepository means git ran but the path is not inside a work tree.
	//
	// The registry refuses to add a directory with no .git entry
	// (internal/project, ErrNotRepository), so the reachable case is a
	// project that stopped being a repository after it was registered — a
	// .git removed by hand, a network volume that unmounted — not a
	// mistake at add time.
	NotARepository Availability = "not-a-repository"
)

// Branch is where the working tree's current branch sits.
type Branch struct {
	// Name is the checked-out branch, or "" when Detached is true.
	Name string `json:"name"`

	// Upstream is the tracking branch's name ("origin/main"), or "" when the
	// branch tracks nothing. Ahead and Behind are only meaningful when it is
	// set — git reports no counts without an upstream to count against.
	Upstream string `json:"upstream"`

	// Ahead and Behind are commits on each side of the upstream.
	Ahead  int `json:"ahead"`
	Behind int `json:"behind"`

	// Detached is HEAD pointing at a commit rather than a branch.
	Detached bool `json:"detached"`

	// Unborn is a repository with no commits yet: `git init` with nothing
	// committed. HEAD names a branch that does not exist, so Name is set
	// while there is nothing for it to point at.
	Unborn bool `json:"unborn"`
}

// FileStatus is one path git reports as differing.
//
// Staged and Worktree are the two halves of porcelain v2's XY pair, kept
// apart rather than collapsed into one badge because the changes panel groups
// by exactly this distinction (DESIGN.md §7): a path can be staged and then
// edited again, which is one entry in both groups.
type FileStatus struct {
	// Path is root-relative and slash-separated — the form the file tree and
	// the editor already use, so a status entry addresses the same rows they
	// do without translation.
	Path string `json:"path"`

	// Staged is how the index differs from HEAD, "" when it does not.
	Staged State `json:"staged"`

	// Worktree is how the working tree differs from the index, "" when it
	// does not.
	Worktree State `json:"worktree"`

	// Conflicted marks an unmerged path. Staged and Worktree are both ""
	// for one: an unmerged path has no index-versus-worktree split to
	// report — it has up to three competing versions — and presenting one
	// would claim a resolution the user has not made.
	Conflicted bool `json:"conflicted"`

	// OrigPath is where a renamed or copied path came from, "" otherwise.
	OrigPath string `json:"origPath"`
}

// Status is one working tree's git state, as of one invocation.
type Status struct {
	Availability Availability `json:"availability"`
	Branch       Branch       `json:"branch"`

	// Files is every path git reported, in git's own order. It is never nil,
	// including on the unavailable paths: it crosses the bridge as JSON, and
	// a null there would reach TypeScript as a value with no length.
	Files []FileStatus `json:"files"`
}

// statusFor builds an empty Status carrying one availability. Files is set
// rather than left nil for the reason the field documents: it marshals to
// JSON, and null would reach TypeScript as a value with no length.
func statusFor(availability Availability) Status {
	return Status{Availability: availability, Files: []FileStatus{}}
}
